### DISCUSSION TITLE
Reached heap limit Allocation failed

### DISCUSSION THREAD

**Discussion by wasims1** [directus/directus#14801] [Archive]
### Describe the Bug

We use `SEARCH` method of directus to send a large request with thousands of data points (the most efficient way to search for a large number or items in one call). However, recently we are getting this error:

```<--- JS stacktrace --->

FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
npm ERR! path /directus
npm ERR! command failed
npm ERR! signal SIGABRT
npm ERR! command sh -c directus "start"

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/node/.npm/_logs/2022-08-02T15_30_42_759Z-debug.log
```

This error is likely due to nodejs memory limit of 2GB. I tried to search directus docs on how to change the memory limit however I couldn't find any. 

How can the memory limits of directus/nodejs be changed? 

Thanks

### To Reproduce

Any large enough request would suffice

### Errors Shown

<--- JS stacktrace --->

FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
npm ERR! path /directus
npm ERR! command failed
npm ERR! signal SIGABRT
npm ERR! command sh -c directus "start"

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/node/.npm/_logs/2022-08-02T15_30_42_759Z-debug.log

### What version of Directus are you using?

9.3.0

### What version of Node.js are you using?

16.13.1

### What database are you using?

mysql

### What browser are you using?

not-relevant

### How are you deploying Directus?

Docker/k8s

**Comment by rijkvanzanten** [comment]
The `--max-old-space-size` flag in NodeJS controls the available memory. The default is only 512mb if I'm not mistaken. Try starting Directus with `NODE_OPTIONS="--max_old_space_size=8192" directus start`. (the 8192 ~8GB of ram. Value is in MiB.). If that doesn't do the trick, you might've ran into a bug in the platform, but I can't tell from the given info 👍🏻

**Reply by wasims1** [reply]
Thanks @rijkvanzanten . Couple updates to everyone ending up on this thread with similar problem:

- default node (depending on version i guess but currently using 16) is 2g

- To get your current heap, this is the command: 
node -e 'console.log(v8.getHeapStatistics().heap_size_limit/(1024*1024))'

- the suggested command: npx --NODE_OPTIONS="--max_old_space_size=8192" directus start 
did not work for us (however we were overriding the directus docker command through k8s which should usually work). I had to manually set the heap using this command: 
export NODE_OPTIONS="--max-old-space-size=8192" 
(you can set the memory size you need)

**Comment by Voldemorten** [comment]
It which terms is "SEARCH" more efficient, than using filters? 
The documentation says: 
> The search parameter allows you to perform a search on all string and text type fields within a collection. It's an easy way to search for an item without creating complex field filters – though it is far less optimized. It only searches the root item's fields, related item fields are not included.

**Reply by rijkvanzanten** [reply]
`search` does a `%LIKE%` search on all textual columns, whereas filters allow you to construct a complex `WHERE` clause. Different tools for different use cases 🙂
